src/core/typed.rs

```diff
@@ -30,6 +32,37 @@ pub struct TypedNode<Witness, App: Application> {
     pub source_ty: Arc<FinalType>,
     /// Target type of the node
     pub target_ty: Arc<FinalType>,
+    /// Index of the node inside the surrounding program
+    pub index: usize,
```

Comment:
In 1587bf728038f6dcf28964d25d9e7be609fd07e7:

It doesn't make sense to me that the index would be in a `TypedNode`. You don't know the indices until after pruning. You say in the commit message that `TypedNode` is "for commitment time" but at commitment time the indices are not known.

The only reason this works is that the current codebase only typechecks programs after they've been linearized, which doesn't seem like a design choice we want to commit to. And the "indices" that are available in an unpruned program are meaningless and (as near as I can tell) never actually used.

Reply:
We will need to come back to that once we implement pruning: Both sharing and pruning may change indices. However, pruning involves a pseudo run of the Bit Machine in order to determine used branches, and the Bit Machine uses `.index` during its execution. This means that we might need `.index` in `TypedNode` after all.

Reply:
Pruning may also change node types, which influences the IMR and static bounds, and it may lead to more sharing. We might want to prune before sharing, if the larger program is not too slow to do so.

Reply:
Interesting observation that the Bit Machine uses `.index`. Probably it could use `get_left` and `get_right` instead?

Also good observation that pruning may change sharing. This strengthens my hope that we can move sharing to the de/serialization layer and forget about indices in the "real" code.